lib/ads/media_tailor_ad_manager.js

```diff
@@ -0,0 +1,860 @@
+            ad.startTimeInSeconds >
+            (previousAd.startTimeInSeconds + previousAd.durationInSeconds)) {
+          previousAd = ad;
+          previousAdBreak = adbreak;
+          break;
+        }
+      }
+    }
+
+    // The cue point gets marked as 'played' as soon as the playhead hits it
+    // (at the start of an ad), so when we come back to this method as a result
+    // of seeking back to the user-selected time, the 'played' flag will be set.
+    if (previousAdBreak && previousAd &&
+        !this.playedAds_.includes(previousAd.adId)) {
+      shaka.log.info('Seeking back to the start of the ad break at ' +
+          previousAdBreak.startTimeInSeconds +
+          ' and will return to ' + currentTime);
+      this.snapForwardTime_ = currentTime;
+      this.video_.currentTime = previousAdBreak.startTimeInSeconds;
+    }
+  }
+
+  /**
+   * @private
+   */
+  onAdBreakEnded_() {
+    const currentTime = this.video_.currentTime;
+    // If the ad break was a result of snapping back (a user seeked over
+    // an ad break and was returned to it), seek forward to the point,
+    // originally chosen by the user.
+    if (this.snapForwardTime_ && this.snapForwardTime_ > currentTime) {
+      this.video_.currentTime = this.snapForwardTime_;
+    }
+    this.snapForwardTime_ = null;
+  }
+
+  /**
+   * @private
+   */
+  onTimeupdate_() {
+    if (!this.video_.duration) {
+      // Can't play yet.  Ignore.
+      return;
+    }
+    if (!this.ad_ && !this.adBreaks_.length) {
+      // No ads
+      return;
+    }
+    const currentTime = this.video_.currentTime;
+    let previousAd = false;
+    if (this.ad_) {
+      previousAd = true;
+      const firstQuartileTime = this.mediaTailorAd_.startTimeInSeconds +
+          0.25 * this.mediaTailorAd_.durationInSeconds;
+      const midpointTime = this.mediaTailorAd_.startTimeInSeconds +
+          0.5 * this.mediaTailorAd_.durationInSeconds;
+      const thirdQuartileTime = this.mediaTailorAd_.startTimeInSeconds +
+          0.75 * this.mediaTailorAd_.durationInSeconds;
+      if (!this.eventsSent.includes('firstQuartile') &&
+          currentTime >= firstQuartileTime) {
+        this.eventsSent.push('firstQuartile');
+        this.sendTrackingEvent_('firstQuartile');
+      } else if (!this.eventsSent.includes('midpoint') &&
+          currentTime >= midpointTime) {
+        this.eventsSent.push('midpoint');
+        this.sendTrackingEvent_('midpoint');
+      } else if (!this.eventsSent.includes('thirdQuartile') &&
+          currentTime >= thirdQuartileTime) {
+        this.eventsSent.push('thirdQuartile');
+        this.sendTrackingEvent_('thirdQuartile');
+      }
+      const remainingTime = this.ad_.getRemainingTime();
+      const duration = this.ad_.getDuration();
+      if (duration > 0 && (remainingTime <= 0 || remainingTime > duration)) {
+        this.onEnded_();
+      }
+      if (this.ad_&& this.ad_.isLinear()) {
+        return;
```

Comment:
Putting an early return statement into the middle of a really long function like this is a bad idea, as far as readability is concerned.
For clarity, this function should probably be broken into multiple sub-functions.

Reply:
done!